src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/ProjectImports/ImportTreeProvider.cs

```diff
@@ -270,15 +271,39 @@ IProjectTree2 SyncNode(IReadOnlyList<IProjectImportSnapshot> imports, IProjectTr
 
                                         if (!ReferenceEquals(child, newChild))
                                         {
-                                            tree = ReplaceChild(child, newChild);
+                                            node = ReplaceChild(child, newChild);
                                         }
                                     }
                                 }
 
-                                return tree;
+                                return node;
+
+                                IProjectTree2 AddChild(IProjectTree2 child) => (IProjectTree2)node.Add(child).Parent!;
+                                IProjectTree2 ReplaceChild(IProjectTree2 oldChild, IProjectTree2 newChild) => (IProjectTree2)node.Remove(oldChild).Add(newChild).Parent!;
+
+                                Dictionary<string, string> GetCaptionByProjectPath()
+                                {
+                                    var result = imports.ToDictionary(i => i.ProjectPath, i => Path.GetFileName(i.ProjectPath));
+
+                                    var isDuplicateByFileName = new Dictionary<string, bool>(StringComparers.Paths);
 
-                                IProjectTree2 AddChild(IProjectTree2 child) => (IProjectTree2)tree.Add(child).Parent!;
-                                IProjectTree2 ReplaceChild(IProjectTree2 oldChild, IProjectTree2 newChild) => (IProjectTree2)tree.Remove(oldChild).Add(newChild).Parent!;
+                                    foreach ((string _, string fileName) in result)
+                                    {
+                                        isDuplicateByFileName[fileName] = isDuplicateByFileName.ContainsKey(fileName);
+                                    }
+
+                                    foreach (IProjectImportSnapshot import in imports)
+                                    {
+                                        string fileName = result[import.ProjectPath];
+                                        
+                                        if (isDuplicateByFileName[fileName])
+                                        {
+                                            result[import.ProjectPath] = $"{fileName} ({Path.GetDirectoryName(import.ProjectPath)})";
+                                        }
+                                    }
+
+                                    return result;
+                                }
```

Comment:
This implementation contains a lot of looping (3 full loops over `imports`, and many loops over `isDuplicateByFileName` via `ContainsKey`) and creates 2 dictionaries in the process. Instead, this works if you don't require the `StringComparers.Paths`:
```suggestion
                                Dictionary<string, string> GetCaptionByProjectPath()
                                {
                                    var result = new Dictionary<string, string>();

                                    foreach (IProjectImportSnapshot import in imports)
                                    {
                                        string fileName = Path.GetFileName(import.ProjectPath);
                                        
                                        result[import.ProjectPath] = result.ContainsValue(fileName) ?
                                          $"{fileName} ({Path.GetDirectoryName(import.ProjectPath)})" :
                                          fileName;
                                    }

                                    return result;
                                }
```

If you do require `StringComparers.Paths`, then the `ContainsValue` needs expanded as it has no way to provide a value comparer:
```suggestion
                                Dictionary<string, string> GetCaptionByProjectPath()
                                {
                                    var result = new Dictionary<string, string>();

                                    foreach (IProjectImportSnapshot import in imports)
                                    {
                                        string fileName = Path.GetFileName(import.ProjectPath);
                                        string caption = fileName;

                                        foreach (string value in result.Values)
                                        {
                                            if (value.Equals(fileName, StringComparers.Paths))
                                            {
                                                caption = $"{fileName} ({Path.GetDirectoryName(import.ProjectPath)})";
                                                break;
                                            }
                                        }

                                        result[import.ProjectPath] = caption;
                                    }

                                    return result;
                                }
```

Reply:
This is not equivalent. If we have `A/B/C` and `X/Y/C` in that order, we need to display `C (A/B)` and `C (X/Y)`. The above code shows `C` and `C (X/Y)`.

> This implementation contains a lot of looping (3 full loops over imports, and many loops over isDuplicateByFileName via ContainsKey) and creates 2 dictionaries in the process. Instead, this works if you don't require the StringComparers.Paths:

Minor point: `ContainsKey` is O(1) not O(N). There's no looping hidden in there.

The tree tends to be fairly small at any given level. I am not massively concerned about the perf impact here.

I'll take another look with a fresh pair of eyes and see if I can make this allocate less. We may be able to reuse the temporary collections as we walk the tree, instead of throwing them away each time.

Reply:
Ah, I understand. It wasn't obvious when reading the original code that both needed to be updated. In that case, you'd just have to update the original caption. The code gets messier so I understand not going this route:
```suggestion
                                Dictionary<string, string> GetCaptionByProjectPath()
                                {
                                    var hasAddedCaption = new List<string>();
                                    var result = new Dictionary<string, string>();

                                    foreach (IProjectImportSnapshot import in imports)
                                    {
                                        string fileName = Path.GetFileName(import.ProjectPath);
                                        string caption = filename;

                                        for (int i = 0; i < result.Keys.Count; ++i)
                                        {
                                            string resultProjectPath = result.Keys[i];
                                            if (result[resultProjectPath].Equals(fileName, StringComparers.Paths))
                                            {
                                                caption = CreateCaption(fileName, import.ProjectPath);
                                                hasAddedCaption.Add(import.ProjectPath);

                                                // Update the original caption since a duplicate has been found.
                                                if (!hasAddedCaption.Contains(resultProjectPath))
                                                {
                                                    result[resultProjectPath] = CreateCaption(result[resultProjectPath], resultProjectPath);
                                                    hasAddedCaption.Add(resultProjectPath);
                                                    break;
                                                }
                                            }
                                        }

                                        result[import.ProjectPath] = caption;
                                    }

                                    return result;

                                    static string CreateCaption(string fn, string pp) => $"{fn} ({Path.GetDirectoryName(pp)})";
                                }
```

Reply:
This now has nested loops, which gives me more concern regarding performance.